bootstrapvz/plugins/admin_user/tasks.py

```diff
@@ -42,13 +43,62 @@ def run(cls, info):
+                        log_check_call(['chroot', info.root, 'chpasswd'],
+                                       info.manifest.plugins['admin_user']['username'] +
+                                       ':' + info.manifest.plugins['admin_user']['password'])
+                        return
+
+                getcreds_path = os.path.join(info.root, 'etc/init.d/ec2-get-credentials')
+                if 'pubkey' in info.manifest.plugins['admin_user']:
+                        import stat
+                        from shutil import copy
+                        full_path = info.manifest.plugins['admin_user']['pubkey']
+                        if not os.path.exists(full_path):
+                                msg = 'Could not find public key at {full_path}'.format(full_path=full_path)
+                                raise TaskError(msg)
+                        log.debug('Copying public key from {path}'.format(path=full_path))
+
+                        if os.path.exists(getcreds_path):
+                                log.warn('You are using a static public key for the admin account.'
+                                         ' This will conflict with the ec2 public key njection mechanisn.'
+                                         ' The ec2-get-credentials startup script has therefore been disabled.')
+                                log_check_call(['chroot', info.root, 'insserv', '--remove',
+                                                'ec2-get-credentials'])
+                        username = info.manifest.plugins['admin_user']['username']
+
+                        ssh_file = os.path.join('/home/'
+                                                '{username}/.ssh/authorized_keys'.format(username=username))
+                        rel_ssh_file = os.path.realpath(info.root + '/{ssh_file}'.format(ssh_file=ssh_file))
+
+                        ssh_dir = os.path.dirname(ssh_file)
+                        rel_ssh_dir = os.path.realpath(info.root + '/{ssh_dir}'.format(ssh_dir=ssh_dir))
+                        if not os.path.exists(rel_ssh_dir):
+                                log.debug('Creating {ssh_dir} mode 700'.format(ssh_dir=rel_ssh_dir))
+                                os.mkdir(rel_ssh_dir, 0700)
+                        else:
+				log.debug('setting {ssh_dir} mode 700'.format(ssh_dir=rel_ssh_dir))
+                                os.chmod(rel_ssh_dir, 0700)
+                        copy(full_path, rel_ssh_file)
+                        mode = (stat.S_IRUSR | stat.S_IWUSR)
+                        os.chmod(rel_ssh_file, mode)
+                        log_check_call(['chroot', info.root, 'chown', '-R', username, ssh_dir])
+                        return
+
+                log.debug('Updating EC2 get credentials script.')
+                username = info.manifest.plugins['admin_user']['username']
+                sed_i(getcreds_path, 'username=\'root\'', 'username=\'{username}\''.format(username=username))
```

Comment:
Should of course only run when the script is actually installed. Probably belongs in a separate task.